### PR TITLE
matmul performance benchmarks

### LINKED ISSUE #44634: torch.sparse improvements - tracking issue
This tracking is meant to list the main improvements we'll be focusing on for sparse tensors, and then mapping them to the releases they should land in. _Please note: these are estimates and indicate current prioritization and require effort - don't rely on this, things may still shift around._

The fill-value proposal (https://github.com/pytorch/rfcs/pull/8) has a good description of the different interpretations of sparse tensors: linear algebra (where unspecified values are assumed to be 0), graph algorithms (where sparse structure indicates nodes/edges), and calculus (where non-zero fill values are needed). _The description of application-level use cases for each of those interpretations is still TODO_.

**~1.7~ 1.8 release**

Sparse documentation updates:

- [x] gh-44635: review and update all content on https://pytorch.org/docs/stable/sparse.html --> PR gh-44635
- [ ] Then update the _"This API is in beta, ..."_ warning at the top of the docs. Feature classification may still be "beta", but API should be stable. Add performance-related note instead.
- [ ] Add an [overview using tables](https://github.com/Quansight-Labs/rfcs/blob/pearu/rfc0005/RFC0003-sparse-roadmap/SparseSupportState.md#nnactivation-functions) of _all_ operators in PyTorch, and show whether they support sparse tensors on CPU, on CUDA, and have autograd support.

**1.8 release**

- [ ] CSR/GCS format (see ~https://github.com/pytorch/rfcs/pull/4, gh-44190~, https://github.com/pytorch/rfcs/pull/13, gh-50937, )
- [x] Performance benchmarks for matrix-matrix and matrix-vector ops (dense-sparse, sparse-sparse, and compare to dense-dense) (see gh-51647)
- [x] Benchmarking framework for sparse that's easy to extend when sparse support is added to new operators (see gh-48397)
- [ ] Sparse test infrastructure improvements (see [here for details](https://github.com/pytorch/pytorch/pull/43330#issuecomment-691210906)) - PR gh-49158
- [ ] Add sparse tensor support for all element-wise functions that map 0 to 0 (this is low-hanging fruit)
- [x] https://github.com/pytorch/pytorch/issues/45113
- [ ] https://github.com/pytorch/pytorch/issues/45897
- [ ] https://github.com/pytorch/pytorch/issues/45996
- [ ] https://github.com/pytorch/pytorch/issues/14945
- [ ] Create public API to replace use of `_indices`, `_values`, `_nnz`. See gh-45695.

**1.9 release**

- [ ] fill-value support in COO and GCS sparse tensors (see https://github.com/pytorch/rfcs/pull/8)
- [ ] extend fill-value support to all functions currently supporting sparse

**To (re)assess**

- [ ] Block sparse tensors, gh-9222

**Relevant but blocked**

- [ ] Sparse-Dense elementwise Multiplication, gh-3158

cc @ezyang @gchanan @zou3519 @vincentqb @aocsa @nikitaved @pearu @v0dro @mruberry

### PR DESCRIPTION
**Pull request by aocsa**
Minor PR following up the previous PR about sparse benchmarking utils https://github.com/pytorch/pytorch/pull/48397

Fixes #44634:  Performance benchmarks for matrix-matrix and matrix-vector ops (dense-sparse, sparse-sparse, and compare to dense-dense)

I ran  all benchmarks on an 2xRTX8000 machine with AMD 2970WX 24-cores for `DLMC/magnitude_pruning` dataset with different sparsity levels.  

---
<details><summary> forward tests (expand for details).
</summary>

- `sparse@sparse`
```
[------------------------------- cpu:matmul-forward -------------------------------]
                           |   0.5   |   0.7   |   0.8   |   0.9   |  0.95  |   0.98
1 threads: -------------------------------------------------------------------------
      torch:dense@dense    |  108.1  |  100.5  |  101.3  |  108.4  |  98.4  |  187.4
      torch:sparse@sparse  |  659.1  |  368.8  |  156.5  |   53.3  |  26.8  |   14.9
      scipy:sparse@sparse  |  565.1  |  233.9  |  130.2  |   23.1  |  21.6  |   15.2

Times are in milliseconds (ms).

[----------------------------------- cuda:matmul-forward -----------------------------------]
                           |    0.5    |    0.7    |   0.8    |   0.9    |   0.95   |   0.98 
1 threads: ----------------------------------------------------------------------------------
      torch:dense@dense    |   2243.5  |   4392.5  |  4419.8  |  2272.3  |  4433.9  |  8920.1
      torch:sparse@sparse  |  21369.2  |  11877.6  |  7339.2  |  1787.2  |  1335.1  |   845.7

Times are in microseconds (us).

```
- `sparse@dense`
```
[------------------------------- cpu:matmul-forward -------------------------------]
                          |   0.5   |   0.7   |   0.8   |   0.9   |   0.95  |   0.98
1 threads: -------------------------------------------------------------------------
      torch:dense@dense   |  105.8  |  103.8  |  103.0  |  104.4  |  104.4  |  197.0
      torch:sparse@dense  |  119.9  |  102.4  |   84.0  |   19.7  |   16.8  |   11.6
      scipy:sparse@dense  |  906.5  |  799.6  |  697.8  |  182.2  |  165.5  |  135.4

Times are in milliseconds (ms).

[------------------------- cuda:matmul-forward --------------------------]
                          |  0.5  |  0.7  |  0.8  |  0.9  |  0.95  |  0.98
1 threads: ---------------------------------------------------------------
      torch:dense@dense   |  2.2  |  4.4  |  4.4  |  2.3  |  4.5   |  2.3 
      torch:sparse@dense  |  5.7  |  6.6  |  4.5  |  1.4  |  1.4   |  1.3 

Times are in milliseconds (ms).

```
- `sparse@vector`
```
[----------------------------------- cpu:matmul-forward ----------------------------------]
                           |    0.5    |   0.7    |   0.8    |   0.9    |   0.95   |   0.98
1 threads: --------------------------------------------------------------------------------
      torch:dense@vector   |    510.6  |   505.8  |   759.6  |   782.1  |   682.4  |  764.6
      torch:sparse@vector  |  10122.8  |  6241.1  |  7935.6  |  2076.3  |  1049.5  |  826.3
      scipy:sparse@vector  |   1756.7  |  1033.9  |   678.2  |   343.5  |   168.5  |   65.4

Times are in microseconds (us).

[-------------------------------- cuda:matmul-forward --------------------------------]
                           |   0.5    |   0.7    |   0.8   |   0.9   |   0.95  |   0.98
1 threads: ----------------------------------------------------------------------------
      torch:dense@vector   |    36.1  |    21.5  |   21.6  |   21.5  |   21.6  |   21.5
      torch:sparse@vector  |  1099.2  |  1289.4  |  775.7  |  327.1  |  285.4  |  274.0

Times are in microseconds (us).

```
</details>

---
<details><summary> backward tests (expand for details).
</summary>

- `sparse@sparse`
```
[--------------------------------- cpu:matmul-backward ---------------------------------]
                           |   0.5    |   0.7    |   0.8    |   0.9    |   0.95  |   0.98
1 threads: ------------------------------------------------------------------------------
      torch:dense@dense    |   246.1  |   315.0  |   306.9  |   168.6  |  290.6  |  146.9
      torch:sparse@sparse  |  6417.5  |  4393.7  |  3012.7  |  1029.4  |  908.0  |  650.7

Times are in microseconds (us).

[----------------------------- cuda:matmul-backward -----------------------------]
                           |   0.5   |   0.7   |   0.8   |  0.9   |  0.95  |  0.98
1 threads: -----------------------------------------------------------------------
      torch:dense@dense    |    6.7  |   13.3  |   13.3  |   6.9  |  13.5  |   6.9
      torch:sparse@sparse  |  143.7  |  143.4  |  119.6  |  29.5  |  29.1  |  10.9

Times are in microseconds (us).

```
- `sparse@dense`
```
 [------------------------------ cpu:matmul-backward -------------------------------]
                          |   0.5   |   0.7   |   0.8   |   0.9   |   0.95  |   0.98
1 threads: -------------------------------------------------------------------------
      torch:dense@dense   |  185.9  |  304.8  |  305.8  |  169.9  |  308.7  |  168.4
      torch:sparse@dense  |  407.9  |  345.8  |  274.6  |  114.2  |  163.6  |  230.5

Times are in milliseconds (ms).

[--------------------------- cuda:matmul-backward --------------------------]
                          |  0.5   |  0.7   |  0.8   |  0.9  |  0.95  |  0.98
1 threads: ------------------------------------------------------------------
      torch:dense@dense   |   6.7  |  13.3  |  13.3  |  6.9  |  13.4  |   6.9
      torch:sparse@dense  |  16.7  |  19.0  |  15.1  |  6.3  |   8.2  |  12.7

Times are in milliseconds (ms).

```  
</details>

Kindly review this  PR. cc @mruberry, @ngimel

### CODE REVIEW

File: benchmarks/sparse/dlmc/spmm.py [hunk 1]
```diff
@@ -0,0 +1,150 @@
...
+            env=device,
+        ) for hidden_size in [512]
+        for sparsity in [0.5, 0.7, 0.8, 0.9, 0.95, 0.98]
+        for label, device, sub_label, stmt in tasks
+        for x, y in load_spmm_dataset(dataset_path, hidden_size, sparsity, spmm_type, device)
+    ]
+    measurements = []
+
+    for i, timer in enumerate(timers * repeats):
+        m = timer.blocked_autorange(min_run_time=0.05)
```
Comment (robieta) [inline benchmarks/sparse/dlmc/spmm.py:126]: Is this enough to get stable measurements?

File: benchmarks/sparse/dlmc/spmm.py [hunk 2]
```diff
@@ -0,0 +1,150 @@
...
+        serialized_results.append(pickle.dumps(m))
```
Comment (robieta) [inline benchmarks/sparse/dlmc/spmm.py:127]: Why do we need to pickle the intermediate results?

File: benchmarks/sparse/dlmc/spmv.py [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+    path = Path()
+    parser = argparse.ArgumentParser(description='spmv Bench')
+
+    parser.add_argument('--path', type=str, help='dataset path')
+    parser.add_argument('--dataset',
+                        type=str,
+                        help='dataset name (random_pruning, variational_dropout, magnitude_pruning, \
+                            extended_magnitude_pruning)',
+                        default='random_pruning')
+    parser.add_argument('--operation',
```
Comment (robieta) [inline benchmarks/sparse/dlmc/spmv.py:28]: This appears to be unused.

**Review by robieta — CHANGES_REQUESTED** [review]
At a high level it seems like `spmm.py` and `spmv.py` could be merged by having a central definition of tasks
```
_TASKS = {
    "mm": (
        ("torch", "torch.mm(dx, dy)"),
        ("torch.sparse", "torch.sparse.mm(x, y)"),
        ("scipy", "scipy_coo_matmul(sx, sy)"),
    ),
    "mv": (
        ("torch", "torch.matmul(dx, vector)"),
        ("torch.sparse", "torch.matmul(x, vector)"),
        ("scipy", "scipy_coo_matmul(sx, vector)"),
    ),
    "backward": (
        ("torch", "torch_backward(dx, dy)"),
        ("torch.sparse", "sparse_torch_backward(x, y)"),
    )
}
```

and some switching code at the start:
```
tasks = _TASKS[operation]
if device != "cpu":
    # SciPy is CPU only.
    tasks = tuple(t for t in tasks if t[0] != "scipy")

loader = (
    load_spmv_dataset if operation == "mv" else
    functools.partial(load_spmm_dataset, spmm_type=spmm_type)
)
```

There's also a bit of logic for globals, but it doesn't seem insurmountable. WDYT?

File: benchmarks/sparse/dlmc/README.md [hunk 4]
```diff
@@ -0,0 +1,20 @@
+# Sparse benchmarks
+
+These sets of benchmarks are for the sparse matrix functionality using a popular real dataset collection called
```
Comment (mruberry) [inline benchmarks/sparse/dlmc/README.md:3]: This README is really cool.

File: benchmarks/sparse/dlmc/spmm.py [hunk 5]
```diff
@@ -0,0 +1,150 @@
...
+import torch
+import torch.utils.benchmark as benchmark_utils
+from .utils import load_spmm_dataset, to_coo_scipy, scipy_coo_matmul
+import pickle
+
+def torch_backward(a_dense, b_dense):
+    a_dense.requires_grad = True
+    b_dense.requires_grad = True
+    r1 = a_dense.matmul(b_dense)
+    f1 = torch.sum(r1)
```
Comment (ngimel) [inline benchmarks/sparse/dlmc/spmm.py:21]: when you are doing `sum` in the function being benchmarked, it alters benchmarking results. You have to pass gradOutput to this function, and then do `r1.backward(gradOutput)`. Same for `sparse_torch_backward` below

File: benchmarks/sparse/dlmc/spmm.py [hunk 6]
```diff
@@ -0,0 +1,150 @@
...
+
+    parser.add_argument('--path', type=str, help='dataset path')
+    parser.add_argument('--dataset',
+                        type=str,
+                        help='dataset name (random_pruning, variational_dropout, magnitude_pruning, \
+                            extended_magnitude_pruning)',
+                        default='random_pruning')
+    parser.add_argument('--operation',
+                        type=str,
+                        help='matmul or backward',
```
Comment (ngimel) [inline benchmarks/sparse/dlmc/spmm.py:50]: It's awkward naming, you should have `matmul` and `matmul_backward` e.g.

File: benchmarks/sparse/dlmc/README.md (hunk 4 above)
Comment (robieta) [reply benchmarks/sparse/dlmc/README.md:3]: Agreed!

File: benchmarks/sparse/dlmc/spmm.py [hunk 7]
```diff
@@ -0,0 +1,150 @@
...
+                ("matmul", device, "torch", "torch.mm(dx, dy)"),
+                ("matmul", device, "torch.sparse", "torch.sparse.mm(x, y)"),
+                ("matmul", device, "scipy", "scipy_coo_matmul(sx, sy)"),
+            ]
+        else:
+            tasks = [
+                ("matmul", device, "torch", "torch.mm(dx, dy)"),
+                ("matmul", device, "torch.sparse", "torch.sparse.mm(x, y)"),
+            ]
+    else:
```
Comment (ngimel) [inline benchmarks/sparse/dlmc/spmm.py:93]: check that it's supported value, otherwise any unexpected `operation` value would be interpreted as `backward`

File: benchmarks/sparse/dlmc/spmv.py [hunk 8]
```diff
@@ -0,0 +1,113 @@
...
+                        type=str,
+                        help='matmul',
+                        default='matmul')
```
Comment (ngimel) [inline benchmarks/sparse/dlmc/spmv.py:31]: you are not providing functions to benchmark backward, is it because backward is not implemented for spmv?

File: benchmarks/sparse/dlmc/spmv.py [hunk 9]
```diff
@@ -0,0 +1,113 @@
...
+    comparison = benchmark_utils.Compare(
+        [pickle.loads(i) for i in serialized_results])
+
+    print("== Unformatted " + "=" * 80 + "\n" + "/" * 95 + "\n")
+    comparison.print()
+
+    print("== Formatted " + "=" * 80 + "\n" + "/" * 93 + "\n")
+    comparison.trim_significant_figures()
+    comparison.colorize()
+    comparison.print()
```
Comment (ngimel) [inline benchmarks/sparse/dlmc/spmv.py:108]: why are you printing both formatted and unformatted versions?

File: benchmarks/sparse/dlmc/spmv.py [hunk 10]
```diff
@@ -0,0 +1,113 @@
...
+        ) for hidden_size in [512]
+        for sparsity in [0.5, 0.7, 0.8, 0.9, 0.95, 0.98]
+        for label, device, sub_label, stmt in tasks
+        for x, y in load_spmv_dataset(dataset_path, hidden_size, sparsity, device)
+    ]
+    measurements = []
+
+    for i, timer in enumerate(timers * repeats):
+        m = timer.blocked_autorange(min_run_time=0.05)
+        serialized_results.append(pickle.dumps(m))
```
Comment (ngimel) [inline benchmarks/sparse/dlmc/spmv.py:90]: As in spmm, no need to pickle intermediates

File: benchmarks/sparse/dlmc/spmv.py [hunk 11]
```diff
@@ -0,0 +1,113 @@
...
+
+    print("== Formatted " + "=" * 80 + "\n" + "/" * 93 + "\n")
+    comparison.trim_significant_figures()
+    comparison.colorize()
+    comparison.print()
+
+    table = [(m.task_spec.sub_label, m.task_spec.description,
+              m.metadata["device"], m.mean) for m in measurements]
+    df = pd.DataFrame(table, columns=['method', 'sparsity', 'device', 'time'])
+    df.to_pickle(df_output_path)
```
Comment (ngimel) [inline benchmarks/sparse/dlmc/spmv.py:113]: why to you want to serialize DataFrame and not comparison? comparison contains the same data. Also, probably disk serialization should be optional.

File: benchmarks/sparse/dlmc/utils.py [hunk 12]
```diff
@@ -0,0 +1,141 @@
...
+def to_coo_scipy(x):
+    indices_1 = x._indices().numpy()
+    values_1 = x._values().numpy()
+    return sparse.coo_matrix((values_1, (indices_1[0], indices_1[1])),
+                             shape=x.shape)
+
+def scipy_coo_matmul(mat1, mat2):
+    if isspmatrix(mat1) and isspmatrix(mat2):
+        result = mat1.dot(mat2).tocoo()
+        return torch.sparse_coo_tensor([result.row, result.col], result.data,
```
Comment (ngimel) [inline benchmarks/sparse/dlmc/utils.py:17]: why are you converting result to torch coo tensor? Would that distort scipy benchmarking results?

File: benchmarks/sparse/dlmc/utils.py [hunk 13]
```diff
@@ -0,0 +1,141 @@
...
+                                       result.shape)
+    else:
+        return mat1.dot(mat2)
```
Comment (ngimel) [inline benchmarks/sparse/dlmc/utils.py:20]: When is this path triggered? For mv? Would mat2 (that is a torch tensor in the mv call) be implicitly converted to ndarray?

File: benchmarks/sparse/dlmc/utils.py [hunk 14]
```diff
@@ -0,0 +1,141 @@
...
+
+
+def read_matrix_params(path):
+    sys.stdin = open(path)
```
Comment (mruberry) [inline benchmarks/sparse/dlmc/utils.py:24]: Should this open the path with a file descriptor and a context manager instead of assigning to sys.stdin?

File: benchmarks/sparse/dlmc/utils.py [hunk 15]
```diff
@@ -0,0 +1,141 @@
...
+    n_rows, n_cols = shape
+    cols = indices
+    rows = [0] * len(cols)
+    for i in range(n_rows):
+        for j in range(indptr[i], indptr[i + 1]):
+            rows[j] = i
+    return torch.tensor([rows, cols], dtype=torch.long)
+
+def load_sparse_matrix(path, device):
+    sys.stdin = open(path)
```
Comment (mruberry) [inline benchmarks/sparse/dlmc/utils.py:38]: Same question here about reading from the path

**Comment by ngimel** [comment]
Do you have results from running this benchmark?

File: benchmarks/sparse/dlmc/spmv.py (hunk 8 above)
Comment (aocsa) [reply benchmarks/sparse/dlmc/spmv.py:31]: Yes, it's seems there is an issue for backward for `torch.matmul` when the input is `spase @ vector`.  Moreover there is no test for this case in the `test_autograd.py`

File: benchmarks/sparse/dlmc/spmm.py (hunk 1 above)
Comment (aocsa) [reply benchmarks/sparse/dlmc/spmm.py:126]: For the maximum problem size, `hidden_size=2048`,  is not enough. I changed it to 1 by default which should be sufficient to be stable and this now can be adjusted in the terminal. 
`    parser.add_argument('--timer_min_run_time', default=1, type=int)`

File: benchmarks/sparse/dlmc/spmm.py (hunk 2 above)
Comment (aocsa) [reply benchmarks/sparse/dlmc/spmm.py:127]: ohhh is not needed, I removed it.  I guessed that by modifying the metadata structure   the `benchmark_utils.Compare` would not work, but actually it works well.

File: benchmarks/sparse/dlmc/spmv.py (hunk 9 above)
Comment (aocsa) [reply benchmarks/sparse/dlmc/spmv.py]: Not needed actually, so I removed the formatted version.

File: benchmarks/sparse/dlmc/utils.py (hunk 12 above)
Comment (aocsa) [reply benchmarks/sparse/dlmc/utils.py]: Not needed actually, I simplified this to `mat1.dot(mat2)`

File: benchmarks/sparse/dlmc/spmv.py (hunk 11 above)
Comment (aocsa) [reply benchmarks/sparse/dlmc/spmv.py]: Not needed actually, I removed this

File: benchmarks/sparse/dlmc/utils.py (hunk 13 above)
Comment (aocsa) [reply benchmarks/sparse/dlmc/utils.py]: All variables are converted to the right format before calling these functions.

**Comment by aocsa** [comment]
> At a high level it seems like `spmm.py` and `spmv.py` could be merged by having a central definition of tasks
> 
> ```
> _TASKS = {
>     "mm": (
>         ("torch", "torch.mm(dx, dy)"),
>         ("torch.sparse", "torch.sparse.mm(x, y)"),
>         ("scipy", "scipy_coo_matmul(sx, sy)"),
>     ),
>     "mv": (
>         ("torch", "torch.matmul(dx, vector)"),
>         ("torch.sparse", "torch.matmul(x, vector)"),
>         ("scipy", "scipy_coo_matmul(sx, vector)"),
>     ),
>     "backward": (
>         ("torch", "torch_backward(dx, dy)"),
>         ("torch.sparse", "sparse_torch_backward(x, y)"),
>     )
> }
> ```
> 
> and some switching code at the start:
> 
> ```
> tasks = _TASKS[operation]
> if device != "cpu":
>     # SciPy is CPU only.
>     tasks = tuple(t for t in tasks if t[0] != "scipy")
> 
> loader = (
>     load_spmv_dataset if operation == "mv" else
>     functools.partial(load_spmm_dataset, spmm_type=spmm_type)
> )
> ```
> 
> There's also a bit of logic for globals, but it doesn't seem insurmountable. WDYT?

Indeed, it was surmountable. Not `spmm.py` and `spmv.py` are merged :)

**Comment by aocsa** [comment]
> Do you have results from running this benchmark?

I addressed the feedback comments. Moreover,  I ran  all benchmarks on an 2xRTX8000 machine with AMD 2970WX 24-cores for `DLMC/magnitude_pruning` dataset with different sparsity levels.  

---
<details><summary> forward tests (expand for details).
</summary>

- `sparse@sparse`
```
[-------------------------------- cpu:matmul-forward -------------------------------]
                           |   0.5   |   0.7   |   0.8   |   0.9   |   0.95  |   0.98
1 threads: --------------------------------------------------------------------------
      torch:dense@dense    |  106.9  |  106.9  |   97.6  |  104.7  |  104.3  |  197.8
      torch:sparse@sparse  |  698.5  |  457.1  |  230.4  |   53.7  |   25.3  |   15.7
      scipy:sparse@sparse  |  553.0  |  221.9  |  114.0  |   22.5  |   18.3  |   11.4

Times are in milliseconds (ms).

[----------------------------------- cuda:matmul-forward -----------------------------------]
                           |    0.5    |    0.7    |   0.8    |   0.9    |   0.95   |   0.98 
1 threads: ----------------------------------------------------------------------------------
      torch:dense@dense    |   2243.5  |   4392.5  |  4419.8  |  2272.3  |  4433.9  |  8920.1
      torch:sparse@sparse  |  21369.2  |  11877.6  |  7339.2  |  1787.2  |  1335.1  |   845.7

Times are in microseconds (us).

```
- `sparse@dense`
```
[------------------------------- cpu:matmul-forward -------------------------------]
                          |   0.5   |   0.7   |   0.8   |   0.9   |   0.95  |   0.98
1 threads: -------------------------------------------------------------------------
      torch:dense@dense   |  105.8  |  103.8  |  103.0  |  104.4  |  104.4  |  197.0
      torch:sparse@dense  |  119.9  |  102.4  |   84.0  |   19.7  |   16.8  |   11.6
      scipy:sparse@dense  |  906.5  |  799.6  |  697.8  |  182.2  |  165.5  |  135.4

Times are in milliseconds (ms).

[------------------------- cuda:matmul-forward --------------------------]
                          |  0.5  |  0.7  |  0.8  |  0.9  |  0.95  |  0.98
1 threads: ---------------------------------------------------------------
      torch:dense@dense   |  2.2  |  4.4  |  4.4  |  2.3  |  4.5   |  2.3 
      torch:sparse@dense  |  5.7  |  6.6  |  4.5  |  1.4  |  1.4   |  1.3 

Times are in milliseconds (ms).

```
- `sparse@vector`
```
[----------------------------------- cpu:matmul-forward ----------------------------------]
                           |    0.5    |   0.7    |   0.8    |   0.9    |   0.95   |   0.98
1 threads: --------------------------------------------------------------------------------
      torch:dense@vector   |    510.6  |   505.8  |   759.6  |   782.1  |   682.4  |  764.6
      torch:sparse@vector  |  10122.8  |  6241.1  |  7935.6  |  2076.3  |  1049.5  |  826.3
      scipy:sparse@vector  |   1756.7  |  1033.9  |   678.2  |   343.5  |   168.5  |   65.4

Times are in microseconds (us).

[-------------------------------- cuda:matmul-forward --------------------------------]
                           |   0.5    |   0.7    |   0.8   |   0.9   |   0.95  |   0.98
1 threads: ----------------------------------------------------------------------------
      torch:dense@vector   |    36.1  |    21.5  |   21.6  |   21.5  |   21.6  |   21.5
      torch:sparse@vector  |  1099.2  |  1289.4  |  775.7  |  327.1  |  285.4  |  274.0

Times are in microseconds (us).

```
</details>


---
<details><summary> backward tests (expand for details).
</summary>

- `sparse@sparse`
```
[--------------------------------- cpu:matmul-backward ---------------------------------]
                           |   0.5    |   0.7    |   0.8    |   0.9    |   0.95  |   0.98
1 threads: ------------------------------------------------------------------------------
      torch:dense@dense    |   246.1  |   315.0  |   306.9  |   168.6  |  290.6  |  146.9
      torch:sparse@sparse  |  6417.5  |  4393.7  |  3012.7  |  1029.4  |  908.0  |  650.7

Times are in microseconds (us).

[----------------------------- cuda:matmul-backward -----------------------------]
                           |   0.5   |   0.7   |   0.8   |  0.9   |  0.95  |  0.98
1 threads: -----------------------------------------------------------------------
      torch:dense@dense    |    6.7  |   13.3  |   13.3  |   6.9  |  13.5  |   6.9
      torch:sparse@sparse  |  143.7  |  143.4  |  119.6  |  29.5  |  29.1  |  10.9

Times are in microseconds (us).

```
- `sparse@dense`
```
 [------------------------------ cpu:matmul-backward -------------------------------]
                          |   0.5   |   0.7   |   0.8   |   0.9   |   0.95  |   0.98
1 threads: -------------------------------------------------------------------------
      torch:dense@dense   |  185.9  |  304.8  |  305.8  |  169.9  |  308.7  |  168.4
      torch:sparse@dense  |  407.9  |  345.8  |  274.6  |  114.2  |  163.6  |  230.5

Times are in milliseconds (ms).

[--------------------------- cuda:matmul-backward --------------------------]
                          |  0.5   |  0.7   |  0.8   |  0.9  |  0.95  |  0.98
1 threads: ------------------------------------------------------------------
      torch:dense@dense   |   6.7  |  13.3  |  13.3  |  6.9  |  13.4  |   6.9
      torch:sparse@dense  |  16.7  |  19.0  |  15.1  |  6.3  |   8.2  |  12.7

Times are in milliseconds (ms).

```  
</details>

cc @robieta, @ngimel, @mruberry

**Review by mruberry — APPROVED** [review]
Cool! Thanks @aocsa. I wonder if we can close the gap with SciPy in the scenarios we're slower on CPU?

@robieta would you like to take another look before this is merged?

**Comment by aocsa** [comment]
> Cool! Thanks @aocsa. I wonder if we can close the gap with SciPy in the scenarios we're slower on CPU?
> 
> @robieta would you like to take another look before this is merged?

Thanks @mruberry, the current implementation of sparse-sparse matmul has almost the same performance.  I forgot to considering the overhead of converting the `csr` output of the `x.dot(y)` scipy operation   to `coo` which is part of the process of `torch.sparse.mm` operation. I updated the results with this change.  

```
[-------------------------------- cpu:matmul-forward -------------------------------]
                           |   0.5   |   0.7   |   0.8   |   0.9   |   0.95  |   0.98
1 threads: --------------------------------------------------------------------------
      torch:dense@dense    |  111.2  |  104.9  |  111.9  |  108.6  |  100.0  |  192.1
      torch:sparse@sparse  |  707.6  |  347.1  |  196.7  |   75.7  |   25.7  |   15.1
      scipy:sparse@sparse  |  584.2  |  365.4  |  105.9  |   30.7  |   22.7  |   14.4
Times are in milliseconds (ms).
```
I moved main comment and benchmark results into the PR description.

**Comment by mruberry** [comment]
Let's give @robieta a chance to review.

File: benchmarks/sparse/dlmc/matmul_bench.py [hunk 16]
```diff
@@ -0,0 +1,125 @@
...
+if __name__ == '__main__':
+    parser = parse_args()
+    args = parser.parse_args()
+
+    if args.with_cuda and not torch.cuda.is_available():
+        raise RuntimeError("No CUDA available")
+
+    dataset_path = args.path
+    dataset_name = args.dataset
+    dataset_path = f"{dataset_path}/{dataset_name}"
```
Comment (robieta) [inline benchmarks/sparse/dlmc/matmul_bench.py:86]: nit: `os.path.join(...)`

**Review by robieta — APPROVED** [review]
I had a single nitpick, but overall looks great. Thanks!

**Comment by aocsa** [comment]
I rebased this PR and I think this  PR is ready to be merged. 
Thanks cc @mruberry

File: test/jit/test_remove_mutation.py [hunk 17]
```diff
@@ -7,7 +7,7 @@
 # Make the helper files in test/ importable
 pytorch_test_dir = os.path.dirname(os.path.dirname(os.path.realpath(__file__)))
 sys.path.append(pytorch_test_dir)
-from torch.testing._internal.jit_utils import JitTestCase, freeze_rng_state	
+from torch.testing._internal.jit_utils import JitTestCase, freeze_rng_state
```
Comment (mruberry) [inline test/jit/test_remove_mutation.py:10]: I'm trying to land this but the internal tools are struggling to reconcile these whitespace changes. Would you revert these changes and/or rebase this PR, @aocsa?

Comment (aocsa) [reply test/jit/test_remove_mutation.py]: Sure, it's done cc @mruberry
